Protect shutdown callbacks lists with a mutex

vadmeste opened the pull request.

vadmeste committed:
Protect shutdown callbacks lists with a mutex

harshavardhana commented on utils.go:
```diff
@@ -90,19 +91,32 @@ var shutdownCallbacks []func() errCode
 
+// Initialize graceful shutdown mechanism
+func prepareGracefulShutdown() {
+	shutdownMutex = &sync.Mutex{}
```
To have a mutex protection for 'callBack' list it is cleaner to put them in a struct together otherwise the mutex relation is harder to justify.

harshavardhana commented on utils.go:
```diff
@@ -90,19 +91,32 @@ var shutdownCallbacks []func() errCode
 type onExitFunc func(code int)
 
+// Initialize graceful shutdown mechanism
+func prepareGracefulShutdown() {
```
Once the latter comment is addressed you can call this function as 'initGracefulShutdown()'

harshavardhana commented on main.go:
```diff
@@ -190,7 +190,7 @@ func main() {
 	// Initialize and monitor shutdown signal
-	shutdownSignal = make(chan bool, 1)
+	prepareGracefulShutdown()
 	monitorShutdownSignal(os.Exit)
```
This function can go potentially inside 'initGracefulShutdown()' - that way everywhere only one function has to be called.

vadmeste committed:
Use data structucre for graceful shutdown

harshavardhana commented on utils.go:
```diff
@@ -77,78 +77,80 @@ func contains(stringList []string, element string) bool {
-// when a shutdown starts. A callback returns 0 for success and 1 for failure.
-// Failure is considered an emergency error that needs an immediate exit
-var shutdownCallbacks []func() errCode
+type ShutdownCallbacks struct {
```
keep shutdownCallbacks as unexported struct.

harshavardhana commented on utils.go:
```diff
@@ -77,78 +77,80 @@ func contains(stringList []string, element string) bool {
-// Protect shutdownCallbacks list from a concurrent access
-var shutdownMutex *sync.Mutex
+// shutdownCBs stores regular and object storages callbacks
+var shutdownCBs ShutdownCallbacks
```
make this a 'pointer here' 

```
var shutdownCBs *shutdownCallbacks
```

harshavardhana committed:
Redo the flow of shutdown.

Introduce a new API on object storage called as Shutdown() which
is implemented per object layer which makes this cleaner.

The reasoning is that registering of callbacks should be done
in a common place and not inside object later initialization.

vadmeste committed:
Merge pull request #1 from harshavardhana/protect_shutdown_mutex-mod

Redo the flow of shutdown.
